tools/codegen/gen.py

```diff
@@ -0,0 +1,1046 @@
+#   - 'model' has the data model for native_functions.yaml.  The classes
+#     in those file represent what you see when you look at
+#     a native_functions.yaml
+#   - 'api' has conversions for how to translate JIT schema into
+#     the various C++ APIs that the codegen interacts with.  There
+#     are in fact THREE different C++ APIs: the public C++ API,
+#     the dispatcher API, and the legacy disaptcher API.  See each
+#     of these respective files for more information
+
+
+# Note [Byte-for-byte compatibility]
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Some special cases we have made in this codegen have been strictly
+# to make sure that git diff -w reports no changes, but we believe
+# they are not semantically meaningful.  After landing the new codegen,
+# we should remove these special cases
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                         HELPER FUNCTIONS
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# Conveniently add error context to exceptions raised.  Lets us
+# easily say that an error occurred while processing a specific
+# context.
+@contextlib.contextmanager
+def context(msg: str) -> Iterator[None]:
+    try:
+        yield
+    except Exception as e:
+        # TODO: this does the wrong thing with KeyError
+        msg = textwrap.indent(msg, '  ')
+        msg = f'{e.args[0]}\n{msg}' if e.args else msg
+        e.args = (msg,) + e.args[1:]
+        raise
+
+# A custom loader for YAML to let us also keep track of line numbers
+# of each entry in the YAML file
+class LineLoader(Loader):
+    def construct_mapping(self, node, deep=False):  # type: ignore
+        mapping = super().construct_mapping(node, deep=deep)  # type: ignore
+        # Add 1 so line numbering starts at 1
+        mapping['__line__'] = node.start_mark.line + 1
+        return mapping
+
+# Parse native_functions.yaml into a sequence of NativeFunctions
+def parse_native_yaml(path: str) -> List[NativeFunction]:
+    with open(path, 'r') as f:
+        es = yaml.load(f, Loader=LineLoader)
+    assert isinstance(es, list)
+    rs: List[NativeFunction] = []
+    for e in es:
+        assert isinstance(e.get('__line__'), int), e
+        loc = Location(path, e['__line__'])
+        funcs = e.get('func')
+        with context(f'in {loc}:\n  {funcs}'):
+            rs.append(NativeFunction.from_yaml(e, loc))
+    return rs
+
+T = TypeVar('T')
+S = TypeVar('S')
+
+# Given a function that operates on NativeFunction, wrap it into a new function
+# that sets some appropriate context managers for that native function.
+# YOU MUST WRAP FUNCTIONS IN THIS for calls to cpp to be sound.
+def with_native_function(func: Callable[[NativeFunction], T]) -> Callable[[NativeFunction], T]:
+    @functools.wraps(func)
+    def wrapper(f: NativeFunction) -> T:
+        with context(f'in {f.loc}:\n  {f.func}'):
+            with local.parametrize(
```

Comment:
I did manual plumbing in the first version of the code. Here is why I decided to switch to dynamic scoping:

1. Manually passed keyword arguments make people less likely to write top-level functions, and encourages locally scoped functions. This is because you have to do write a bunch more boilerplate to pass the argument (`*, use_c10_dispatcher_full: bool` in argument list, and then `use_c10_dispatcher_full=use_c10_dispatcher_full` for all downstream calls). If you put the function locally, you can skip the recall boilerplate, but I also agree with you that less nested functions makes things easier to understand.
2. Passing values by hand adds friction to adding/removing hacks like this. Now, arguably, you *want* friction when you add hacks. But there's already a pretty large pile of hacks, and we also want to make it easy to remove hacks. If you remove a hack, now you have to *un-plumb* the all of the places you plumbed the manually passed argument. If you don't make it easy for people to remove hacks, then people will keep putting it off. Better to make it easy.
3. A minor point, but extra keyword argument impedes common functional programming idioms. If I want to map a function over a list, this is easy to do if the function takes only one argument, but is much more annoying if I also have to pass a keyword argument to it (no currying).

Reply:
Ok fair enough. Per #2 (and the other comment), it'd be nice to do what we can to make the add/remove friction asymmetric :)